### PR TITLE
New operation abort and restart

### PR DESCRIPTION
**Pull request by stefanbrausch**
This pull request introduce the new operation "abort and restart".

Motiviation for the change: Some of our testing jobs have problems with timeouts. In this case the job would be aborted and restarted now.

### CODE REVIEW

File: src/main/java/hudson/plugins/build_timeout/operations/AbortAndRestartOperation.java [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+                hudson.plugins.build_timeout.Messages.Timeout_Aborted())
+        );
+        
+        Executor e = build.getExecutor();
+        if (e != null) {
+            e.interrupt(Result.ABORTED);
+        }
+        
+        if (checkMaxRestarts(build)) {
+            build.getRootBuild().getProject().scheduleBuild(new BuildTimeoutAbortAndRestartCause());
```
Comment (oleg-nenashev) [inline src/main/java/hudson/plugins/build_timeout/operations/AbortAndRestartOperation.java]: What about build parameters

**Comment by oleg-nenashev** [comment]
For me it smells like a duplicate of Naginator's plugin functionality. I'm OK with it, but probably it makes sense to integrate plugins somehow.

File: src/main/java/hudson/plugins/build_timeout/operations/AbortAndRestartOperation.java [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+        // Use messages in hudson.plugins.build_timeout.Messages for historical reason.
+        listener.getLogger().println(hudson.plugins.build_timeout.Messages.Timeout_Message(
+                effectiveTimeoutMinutes,
+                hudson.plugins.build_timeout.Messages.Timeout_Aborted())
+        );
+        
+        Executor e = build.getExecutor();
+        if (e != null) {
+            e.interrupt(Result.ABORTED);
+        }
```
Comment (ikedam) [inline src/main/java/hudson/plugins/build_timeout/operations/AbortAndRestartOperation.java]: It's better to simply call `new AbortOperation().perform`.

**Comment by ikedam** [comment]
You'd be interested in how [Rebuild plugin|https://wiki.jenkins-ci.org/display/JENKINS/Rebuild+Plugin] reruns builds.
* That preserves build parameters.
* That preserves build causes.
** This preserves relation between upstream builds and allows retriggered builds to copy artifacts again.
** Additionally, rebuild plugin displays very useful link "This build is a rebuild of #xxxx".

**Comment by oleg-nenashev** [comment]
@ikedam 
I'd vote for Naginator, but the reason is same. Rebuild seems to be more healthy, so I will agree with such approach as well

**Comment by ikedam** [comment]
I agree that "fail the build" and [Naginator plugin|https://wiki.jenkins-ci.org/display/JENKINS/Naginator+Plugin]  provides the same functionality.
Though that doesn't make this request rejected immediately, it's prefered that this feature is useful than Naginator plugin in some points.

**Comment by oleg-nenashev** [comment]
@ikedam 
I agree, such strategy would be useful. I'm mostly aware about the backend mplementation

File: src/main/java/hudson/plugins/build_timeout/operations/AbortAndRestartOperation.java [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+            
+            count++;
+         
+            if (count >= this.maxRestarts) {
+                return false;
+            }
+            build = build.getPreviousBuild();
+        }
+        return true;
+    }
```
Comment (ikedam) [inline src/main/java/hudson/plugins/build_timeout/operations/AbortAndRestartOperation.java]: Let's record retrying counts with `Cause` or `Action`.
That will be more stable.

Comment (Jochen-A-Fuerbacher) [inline src/main/java/hudson/plugins/build_timeout/operations/AbortAndRestartOperation.java]: @ikedam Do you mean to write retrying counts in Cause of each build? And why exactly is the code above less stable?

Comment (ikedam) [inline src/main/java/hudson/plugins/build_timeout/operations/AbortAndRestartOperation.java]: > Do you mean to write retrying counts in Cause of each build?

Or to link to the build that triggered the current build.

> And why exactly is the code above less stable?

`getPreviousBuild` doesn't always return the last retrying build.
Users can schedule builds while another build is running. That can result wrong retrying counts.

**Comment by stefanbrausch** [comment]
First: The reason why we are use build-timeout and not periodic reincarnation (it's similar to nagiator plugin) are the email notification.

**Comment by stefanbrausch** [comment]
@Jochen-A-Fuerbacher has created a new pull request

**Comment by ikedam** [comment]
JFYI.
You don't need to create new pull requests for new changes.
You can update a pull request by pushing new commits to your branch.

File: src/main/java/hudson/plugins/build_timeout/operations/AbortAndRestartOperation.java [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+    
+    @DataBoundConstructor
+    public AbortAndRestartOperation(int maxRestarts){
+        this.maxRestarts = maxRestarts;
+    }
+       
+    /**
+     * @param build
+     * @param listener
```
Comment (KostyaSha) [inline src/main/java/hudson/plugins/build_timeout/operations/AbortAndRestartOperation.java]: remove it if you have no real description

Comment (KostyaSha) [inline src/main/java/hudson/plugins/build_timeout/operations/AbortAndRestartOperation.java]: ah, sorry, it already closed.

**Pull request closed without merging**